### PR TITLE
VOTE SLEP018 - Pandas Output for Transformers

### PR DESCRIPTION
**Pull request by thomasjpfan**
This PR is for us to discuss and collect votes for [SLEP018 - Pandas Output for Transformers](https://scikit-learn-enhancement-proposals--72.org.readthedocs.build/en/72/slep018/proposal.html). The current implementation is available at https://github.com/scikit-learn/scikit-learn/pull/23734. Note that this vote is for the API and the implementation can be adjusted.

According to our [governance model](https://scikit-learn.org/stable/governance.html#decision-making-process), the vote will be open for a month (till 17th August), and the motion is accepted if 2/3 of the cast votes are in favor.

@scikit-learn/core-devs

### CODE REVIEW

**Comment by adrinjalali** [comment]
I'd say most users don't have a separate pipeline for their transform steps and another pipeline for adding the final predictor. How does a usual pipeline would look like? Should users do sth like `pipeline[:-1].set_output(...)` or can they call `set_output` on the pipeline and expect that to apply only to steps with an available `transform` method?

Otherwise I'm happy with the SLEP.

**Comment by adrinjalali** [comment]
Then the example in the SLEP could also mirror that to make it clear. But it's a +1 for me anyway :)

**Review by adrinjalali — APPROVED** [review]

**Comment by thomasjpfan** [comment]
In this SLEP, I updated the pipeline example to have a classifier showcasing how `set_output` can be called on the whole pipeline and only the transformers are configured.

**Comment by amueller** [comment]
+1

**Comment by jnothman** [comment]
So can I clarify that `Pipeline` is exceptional in the sense that it is the only non-transformer that has a `set_output` method (and that it only affects the output of the Pipeline if either it is also a transformer, or some pipeline components behave differently with different input)?

(Do we have other non-transformers that have a transformer for a parameter, aside from TransformedTargetRegressor?)

**Comment by thomasjpfan** [comment]
> Do we have other non-transformers that have a transformer for a parameter, aside from TransformedTargetRegressor?

`GridSearchCV` can define a [transform method](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html#sklearn.model_selection.GridSearchCV.transform) if the underlying estimator is a transformer.

Thinking about it more, the special case for Pipeline can influence other meta-estimators. For example, a `VotingClassifer` with many pipelines:

```python
voting = VotingClassifier([
    ("pipe1", pipe1), ("pipe2", pipe2), ("pipe3", pipe3)
])

# If `VotingClassifier` defines a `set_output`, then the whole pipeline can be configured with:
voting.set_output(transform="pandas")

# If not, then every pipeline needs to be set individually:
voting2 = VotingClassifier([
    ("pipe1, pipe1.set_output(transform="pandas")), ...
])
```

For a better UX, I think all first-party meta-estimators should define a `set_output` and configures all their inner estimators. Specifically for a meta-estimator:

1. If the inner estimator has a `set_output`, call it (the case of `Pipeline`).
2. If the inner estimator is a transformer, then call `set_output` on it.
3. Otherwise do nothing.

**Comment by ogrisel** [comment]
+1.

**Review by jnothman — APPROVED** [review]

**Review by glemaitre — APPROVED** [review]
+1

**Review by agramfort — APPROVED** [review]

**Review by TomDLT — APPROVED** [review]

**Review by lorentzenchr — APPROVED** [review]

**Comment by amueller** [comment]
ping @GaelVaroquaux ;)

**Review by GaelVaroquaux — APPROVED** [review]
Thanks for the thoughtful SLEP, the discussions and the prototype, @thomasjpfan.

Thanks for the ping, @amueller :)

File: slep018/proposal.rst [hunk 1]
```diff
@@ -37,17 +37,23 @@ sparse data, e.g. ``OneHotEncoder(sparse=True)``, then ``transform`` will raise
...
 
    # X_trans_df is again a pandas DataFrame
-   X_trans_df = num_preprocessor[0].transform(X_df)
+   X_trans_df = log_reg[0].transform(X_df)
+
+   # The classifier contains the feature names in
+   log_reg[-1].feature_names_in_
 
 Meta-estimators that support ``set_output`` are required to configure all inner
 transformer by calling ``set_output``. Specifically all fitted and non-fitted
```
Comment (jjerphan) [inline slep018/proposal.rst:59]:
```suggestion
Meta-estimators that support ``set_output`` are required to configure all inner
transformers by calling ``set_output``. Specifically all fitted and non-fitted
```

File: slep018/proposal.rst [hunk 2]
```diff
@@ -37,17 +37,23 @@ sparse data, e.g. ``OneHotEncoder(sparse=True)``, then ``transform`` will raise
 ``ValueError`` if ``set_output(transform="pandas")``. Dealing with sparse output
 might be the scope of another future SLEP.
```
Comment (jjerphan) [inline slep018/proposal.rst:38]: > Dealing with sparse output might be the scope of another future SLEP.

Nit: Might it be worth mentioning it in the `Discussion` section?

**Review by jjerphan — APPROVED** [review]
+1

Thank you for the heads-up, @thomasjpfan.

Here are some minor nitpicks.

**Review by lesteve — APPROVED** [review]
+1

I pushed a commit with a typo fix: StandardScal**a**r -> StandardScal**e**r

**Comment by thomasjpfan** [comment]
I am also +1 on this SLEP. Including my vote we have 12 in favor and 0 against which means this enhancement proposal is accepted. Thank you everyone for making this possible!

**Pull request merged**